GPU access in the sandbox

SomeoneSerge opened the pull request:
## Description of changes

The PR introduces a way to expose devices and drivers (e.g. GPUs and `/run/opengl-driver/lib`) in the Nix build's sandbox for specially marked derivations. The approach is that described in https://github.com/NixOS/nixpkgs/issues/225912: we introduce a pre-build-hook which would scan the `.drv` being built for the presence of a chosen tag in the `requiredSystemFeatures` attribute and, upon a match, instruct Nix to mount a list of device-related paths. E.g. one could mark the derivation as `mkDerivation { ...; requiredSystemFeatures = [ "cuda" ]; }` to indicate that it would need to use CUDA during the build. Untagged derivations would still observe "pure" environment without any devices. Further, the `requiredSystemFeatures` attribute is special, Nix uses it to determine whether a host is capable of building the derivation, and which remote builders to choose. A host that hasn't been set up with `nix.settings.system-features` (`system-features` in `nix.conf(5)`) would refuse to build the marked derivation

One application of this hook is to implement GPU tests as normal derivations (cf. the torch example in the diff).
Another use could be to use Nix to run ML/DL/NumberCrunching pipelines that utilize GPUs (e.g. https://github.com/SomeoneSerge/pkgs/blob/5bf5ea2b16696e70ade5a68063042dcf2e8f033b/python-packages/by-name/om/omnimotion/package.nix#L270-L291). Similarly, one could write derivations that render things using OpenGL or Vulkan

How to test:

1. Take a NixOS machine with an nvidia GPU and `hardware.opengl.enable = true`.
2. Try building `python3Packages.torch.tests.cudaAvailable`, observe it's refused evaluation
3. Rebulid with `programs.nix-required-mounts.enable = true` (

    <details>

    ```
    # Note: nix-required-sandbox-paths extends system-features,
    # but they won't be merged with the nixos defaults, because they got a different priority
    # -> set manually:
    nix.settings.system-features = [
      "nixos-test"
      "benchmark"
      "big-parallel"
      "kvm"
    ];
    programs.nix-required-sandbox-paths.enable = true;
    ```
   
    </details>
4. Build `python3Packages.torch.gpuChecks.cudaAvailable` successfully
5. Try dropping `cuda` from `gpuChecks.cudaAvailable`'s `requiredSystemFeatures`, and observe the test fail

CC @NixOS/cuda-maintainers 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Alternatives and related work

@tomberek suggests that one could have derivations ask for specific paths using `__impureHostDeps`, currently undocumented and being used for Darwin builds

For GPU tests, @Kiskae suggests one could instead use NixOS tests with the PCI passthrough

We could also possibly consider supporting variables like `CUDA_VISIBLE_DEVICES` and/or re-using nvidia-docker/nvidia-container-toolkit because that's the interface many (singularity, docker) rely on. These currently do not support any static configuration, and in particular they rely on glibc internals to discover the host drivers in a very non-cross-platform way.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SomeoneSerge committed:
nix-required-mounts: init

SomeoneSerge committed:
python3Packages.pynvml: add a gpu test

SomeoneSerge committed:
python3Packages.torch{,-bin}: test torch.cuda.is_available()

SomeoneSerge commented:
A slightly more involved usage example: https://github.com/SomeoneSerge/pkgs/blob/a23cebfa6c183538a89b7aedceccd1702400b9e2/python-packages/by-name/om/omnimotion/package.nix#L270-L288

SomeoneSerge committed:
blender: add CUDA discovery test

SomeoneSerge commented:
- [x] Added a test for blender as suggested by @the-furry-hubofeverything in https://github.com/NixOS/nixpkgs/issues/225912?notification_referrer_id=NT_kwDOAJRS1LI2MTM5OTA3NTk1Ojk3MjA1MzI#issuecomment-1760634358
- [x] Updated the PR description to reflect that the approach can be used (is already being used) to manage generic GPU workloads
- [ ] I'm going to rewrite the structure of the NixOS module and of the hook's config in a bit
- [ ] I'd like to switch to the approach described by @Kiskae in https://github.com/NixOS/nixpkgs/issues/225912?notification_referrer_id=NT_kwDOAJRS1LI2MTM5OTA3NTk1Ojk3MjA1MzI#issuecomment-1762883674. Specifically, I want to expose the GPU checks such that they can be run outside the sandbox (what we use to verify our builds are correct, the user may utilize to verify the host configuration). These checks can then be wrapped to be exposed in `passthru.tests` to run with `requiredSystemFeatures`, or to be run as NixOS tests with PCI passthrough when somebody comes around to implement that

SomeoneSerge committed:
nixosTests.nix-required-mounts: init

SomeoneSerge commented:
- [x] Added a NixOS test verifying that the sandbox is relaxed if and only if the derivation has asked for the respective `requiredSystemFeature`

SomeoneSerge commented on nixos/modules/programs/nix-required-mounts.nix:
```diff
@@ -0,0 +1,83 @@
+  options.programs.nix-required-mounts = {
+    enable = lib.mkEnableOption
+      "Expose extra paths to the sandbox depending on derivations' requiredSystemFeatures";
+    presets.opengl.enable = lib.mkOption {
+      type = lib.types.bool;
+      default = config.hardware.opengl.enable;
+      description = ''
+        Expose OpenGL drivers to derivations marked with requiredSystemFeatures = [ "opengl" ]
+      '';
+    };
+    presets.cuda.enable = lib.mkEnableOption ''
+      Expose CUDA drivers and GPUs to derivations marked with requiredSystemFeatures = [ "cuda" ]
+    '';
```
These are just helpers so people don't have to write `allowedPatterns` manually. Should I move these to `hardware.opengl` and `hardware.nvidia` respectively?

SomeoneSerge commented on pkgs/by-name/ni/nix-required-mounts/package.nix:
```diff
@@ -0,0 +1,57 @@
+} ''
+  ${lib.getExe buildPackages.python3.pkgs.flake8} ${./main.py}
+
+  cat > main.py << EOF
+  #!${lib.getExe python3}
+
+  $(cat ${./main.py})
+  EOF
+
+  sed -ie '
+    /^entrypoint()$/ {
+      x ;
+      r ${confPath}
+      }' main.py
+
+  echo "entrypoint()" >> main.py
```
Fairly ugly, but I didn't want any environment variables/runtime stuff, just a static "resource" included in the ~~binary~~ script. If you suggest a cleaner way I'll happily rewrite this

samuela replied:
This whole thing feels way more complicated than it needs to be... why not just use [makeWrapper](https://nixos.org/manual/nixpkgs/stable/#fun-makeWrapper) with environment variables?

SomeoneSerge replied:
Yea, maybe (Someone just wants `std::embed`)

SomeoneSerge commented on pkgs/by-name/ni/nix-required-mounts/main.py:
```diff
@@ -0,0 +1,121 @@
+            file=stderr,
+        )
+
+    proc = subprocess.run(
+        [
+            CONFIG["nixExe"],
+            "show-derivation",
+            drv_path,
+        ],
+        capture_output=True,
```
IIRC this requires the `nix-command` experimental feature. Should I just assume the 1st argument to always be the path to the `.drv` and read it directly?

SomeoneSerge commented on pkgs/by-name/ni/nix-required-mounts/main.py:
```diff
@@ -0,0 +1,121 @@
+    args = parser.parse_args()
+    drv_path = args.derivation_path
+
+    if not Path(drv_path).exists():
+        print(
+            f"[E] {drv_path} doesn't exist."
+            " This may happen with the remote builds."
+            " Exiting the hook",
+            file=stderr,
+        )
```
When a machine with `programs.nix-required-mounts.enable = true` is set up as a remote builder for another machine, the remote builds do not seem to copy the `.drv` over, at least not by the time the `pre-build-hook` executes. The current solution is to do nothing and not expose any paths

Thoughts?

layus replied:
I think it boils down to this feature/optimization:

When you are privileged (or building a content-addressed derivation), you can build a derivation without uploading all the .drv closure first. This is of course less secure (because you can forge arbitrary output hashes) but faster if the closure contains very large files as inputs. Or if the closure is large. The .drv you send is not trusted, nor trustable, and not written to the builder store. Just used for building and sending back the result.

When you are not privileged (and not building a content-addressed derivation), you need to send the whole closure, and the .drv file will be present, as its validity can be recursively verified. 

See https://github.com/NixOS/nix/blob/d12c614ac75171421844f3706d89913c3d841460/src/build-remote/build-remote.cc#L302-L334 And the referenced comment https://github.com/NixOS/nix/blob/d12c614ac75171421844f3706d89913c3d841460/src/libstore/daemon.cc#L589-L622

layus replied:
It bugged me too much :-D https://github.com/NixOS/nix/issues/9272

SomeoneSerge replied:
Oh gosh, thanks a bunch @layus!

hacker1024 replied:
Also see: https://github.com/nix-community/infra/pull/1807#discussion_r2070205287

samuela commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -225,6 +225,16 @@ stdenv.mkDerivation (finalAttrs: rec {
+      "blender-cuda-available"
+      {
+        nativeBuildInputs = [ finalAttrs.finalPackage ];
+        requiredSystemFeatures = [ "cuda" ];
```
is there a way we could avoid having a stringly-typed API here? eg something like
```suggestion
        requiredSystemFeatures = [ system-features.cuda ];
```
I'm a fan of whatever works, but I think avoiding a stringly-typed interface would be preferable, ceteris paribus.

SomeoneSerge replied:
Actually, let me clarify: I made a preset with `"cuda"` because it seemed like the simplest string I could come up with. I'm wary of introducing other names or more infra around these names, because it's easy to slip down the rabbit hole. One obvious rabbit hole is that you could start making derivations that ask for `requiredSystemFeatures = let formatFeature = capabilities: (concatStringsMapSep "-" (x: "sm_{x}") capabilities); [ (formatFeature cudaCapabilities) ]` and deploying hosts that declare `programs.nix-required-mounts.allowedPatterns.cuda.onFeatures = map allowformatFeature (genSubsets cudaCapabilities)`

SomeoneSerge replied:
The final value, at least for now, would still be a string, but we could define somewhere a dictionary of valid literals and have people reference them rather than write the strings directly. Could be as silly as `cudaPackages.cudaFeature = "cuda"` and `requiredSystemFeatures = [ cudaPackages.cudaFeature ]`.

We then could even extend this to `cudaFeature = f cudaFlags`

SomeoneSerge replied:
But what are your thoughts on modeling the capabilities?

SomeoneSerge replied:
I think `(import <nixpkgs> { config.cudaCapabilities = [ "8.6" ]; }).blender.tests.cuda-available.requiredSystemFeatures = [ "cuda-sm_86" ]` would be really great, but we run into the limitations of this mechanism very soon

SomeoneSerge replied:
(Nix with an external scheduler when)

ConnorBaker replied:
I think we should model capabilities -- they're essentially features of the hardware which determine how different pieces of software should be built and whether or not they can run on the system.

I do think it's a bit out of scope of this PR though; @SomeoneSerge would you make an issue so we can track this and have a longer discussion there?

samuela commented on pkgs/by-name/ni/nix-required-mounts/main.py:
```diff
@@ -0,0 +1,121 @@
+    return out
+
+
+def entrypoint():
+    if __name__ != "__main__":
+        return
```
this pattern is unconventional... why not
```
def entrypoint():
    ...

if __name__ == "__main__":
    entrypoint()
```

SomeoneSerge replied:
Because I wanted this to be called _after_ the redefinition of CONFIG 🤡

samuela replied:
resolving this and switching convo to https://github.com/NixOS/nixpkgs/pull/256230/files#r1362949109

samuela commented on pkgs/by-name/ni/nix-required-mounts/main.py:
```diff
@@ -0,0 +1,121 @@
+
+    patterns = list(chain.from_iterable(allowed_patterns[f]["paths"] for f in features))  # noqa: E501
+
+    roots: List[Path] = []
+
+    for pattern in patterns:
+        roots.extend(Path(path) for path in glob.glob(pattern))
+
+    roots = sorted(set(roots))
```
```suggestion
    roots = sorted(set(Path(path) for pattern in patterns for path in glob.glob(pattern)))
```

SomeoneSerge replied:
Applied manually because `black` formatting

samuela commented on pkgs/by-name/ni/nix-required-mounts/main.py:
```diff
@@ -0,0 +1,121 @@
+    )
+
+    drv_env = drv[canon_drv_path].get("env", {})
+    features = drv_env.get("requiredSystemFeatures", [])
+    if isinstance(features, str):
+        features = features.split()
```
```suggestion
    features = drv_env.get("requiredSystemFeatures", "").split()
```

SomeoneSerge replied:
Are `env.*` values always going to be a string? Whatever happens when one uses structured attrs?

samuela replied:
IIUC they should be exactly equivalent: 
* if `"requiredSystemFeatures"` exists in `drv_env`, then `drv_env.get("requiredSystemFeatures", "").split() => drv_env["requiredSystemFeatures"].split()`
* if `"requiredSystemFeatures"` does not exist in `drv_env`, then `drv_env.get("requiredSystemFeatures", "").split() => "".split() => []`.

SomeoneSerge replied:
What I meant is, I thought structuredAttrs would prevent requiredSystemFeatures the list from collapsing into a string. But I didn't check if the list really goes into the env.*

Kiskae replied:
Maybe refer to the code that nix uses to resolve it?

https://github.com/NixOS/nix/blob/201c115c3e242e3db60a2cd1583223d4e196d6e5/src/libstore/parsed-derivations.cc#L64-L99

If `structuredAttrs`, then it is a array of strings in the "__json" json serialization.
Else it is a whitespace-separated string in the environment.

SomeoneSerge replied:
Thanks @Kiskae! Rewrote the bit in the last commit

samuela reviewed (commented):
you a real og for this one @SomeoneSerge !! 🚀

SomeoneSerge committed:
nix-required-mounts: use wrappers instead of statically embedding config into the script

SomeoneSerge committed:
nix-required-mounts: handle __structuredAttrs

samuela commented on pkgs/development/python-modules/torch/default.nix:
```diff
@@ -4,6 +4,10 @@
 
+  # tests.gpu:
+  callPackage,
+  torch,
```
Nix allows this? Passing `torch` to the derivation that makes `torch`?

SomeoneSerge replied:
Sure, because it's not used by torch itself, only by the passthru

SomeoneSerge replied:
Sure, because it's not used by torch itself, only by the passthru

ConnorBaker commented on nixos/modules/programs/nix-required-mounts.nix:
```diff
@@ -0,0 +1,85 @@
+    '';
+    allowedPatterns = with lib.types;
+      lib.mkOption rec {
+        type = attrsOf patternType;
```
@SomeoneSerge in my experience, `attrsOf` and `listOf` and friends expect their argument to be a type -- I expected to see `patternType.type` here instead.

As an example:

https://github.com/NixOS/nixpkgs/blob/157b25b88712903f31d159db55f56d91331f141f/pkgs/development/cuda-modules/modules/generic/manifests/feature/release.nix#L65

SomeoneSerge replied:
`patternType` is a type, it's an instance of `submodule`

ConnorBaker replied:
Oh! That makes more sense -- thanks!

ConnorBaker reviewed (changes requested):
Mostly looks good! Just a few nits.

I still need to try to run a functional test.

SomeoneSerge commented on pkgs/by-name/ni/nix-required-mounts/package.nix:
```diff
@@ -0,0 +1,47 @@
+{ addOpenGLRunpath
+, cmake
+, allowedPatterns ? rec {
+    opengl.onFeatures = [ "opengl" ];
+    opengl.paths = [
+      addOpenGLRunpath.driverLink
+      "/dev/video*"
+      "/dev/dri"
+    ];
+    cuda.onFeatures = [ "cuda" ];
+    cuda.paths = opengl.paths ++ [
+      "/dev/nvidia*"
+    ];
+  }
```
Indeed. The package and the module end up being different because the module also includes the symlink targets in the nix store, whereas the package in principle could be used on a fhs machine.

But maybe I just put the shared part into the passthru

SomeoneSerge commented on pkgs/development/python-modules/torch/default.nix:
```diff
@@ -4,6 +4,10 @@
 
+  # tests.cudaAvailable:
+  callPackage,
+  torch,
```
No, I'm not sure we want that. If anything, we may want to take specifically `torchWithCuda` from the package set instead of `torch`. It's the same for all other passthru tests that are based on overriding the parent derivation with `doCheck`: they really test the instance from the package set, even if you override it locally. If you want to test the overridden version, I think you have to overlay the package set.

Besides, I'm pretty sure `self = buildPythonpackage` will actually fail at `overridePythonAttrs`. Maybe `finalAttrs` could work, if we really wanted to make this happen. Then we should do the same thing to the rest of nixpkgs:)

SomeoneSerge commented on pkgs/development/python-modules/torch/test-cuda.nix:
```diff
@@ -0,0 +1,22 @@
+} ''
+  python3 << EOF
+  import torch
+  assert torch.cuda.is_available()
+  print(f"{torch.cuda.is_available()=}")
```
I think that's what I did in the beginning and then swapped it because when `is_available()=False` you see the message twice: once from the print, once from assert

ConnorBaker replied:
Ah, I see what you're saying -- that's okay then

SomeoneSerge committed:
programs.nix-required-mounts: inherit defaults from the package

SomeoneSerge committed:
nix-required-mounts: expose the VM test in passthru

SomeoneSerge committed:
nix-required-mounts: link the issue about unavailable .drvs

SomeoneSerge committed:
nix-required-mounts: fix: add missing metadata

SomeoneSerge commented:
Thanks @ConnorBaker!

I force-pushed one last time:

- The "default" `allowedPatterns` has been deduplicated (now it's enough to edit the paterns in the package) as suggested in the reviews
- I also realized that with the package visible at the top-level, it's really not a good thing to omit the metadata, so I added that and ditched the `runCommand` stuff as well
- Linked the issue opened by @layus in the code

I reran `nixosTests.nix-required-mounts` (for reference, the output looks like [so](https://gist.github.com/SomeoneSerge/1c89104caeb09f57a93cf5facd57f93e)), I also rebuilt my NixOS workstation against HEAD and checked `torch.cudaAvailable`.

I'm ready to have this merged

Madouura commented:
Why not just use [``makeImpureTest``](https://github.com/NixOS/nixpkgs/blob/73efbd90a7cb3375ab08a3aee77002e18aecf1e3/pkgs/build-support/make-impure-test.nix)?

SomeoneSerge commented:
> Why not just use [makeImpureTest](https://github.com/NixOS/nixpkgs/blob/73efbd90a7cb3375ab08a3aee77002e18aecf1e3/pkgs/build-support/make-impure-test.nix)?

Thanks @Madouura, I wasn't aware of this prior work!

From what I see the scope is different, but also some implementation choices also differ and happen to be relevant to our use-case:

1. `makeImpureTest` lets a derivation list the additional paths statically, just like `__impureHostDeps` does.
  
   In case of e.g. CUDA tests we also need to expose the `libcuda.so` driver, which we only know the location of for NixOS hosts (and even that might change in later releases), so the user has to provide the actual location of the driver, and it would have to be mapped to `/run/opengl-driver/lib` in the sandbox. Currently, this should still be an issue with mesa as well
2. `makeImpureTest` transforms a derivation into a script, that outputs the correct nix-build invocation for the user manually to pick up and execute (`$(...)`). That's two sequential nix-builds per test and there's a kind of a [red-blue](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) situation about normal versus "impure test" derivations: you can't just assemble several "impure tests" into one derivation or an attrset of derivations that nix-build or other tools would just automatically pick up and build. I think, though, we could in principle rewrite `makeImpureTest` using the `__impureHostDeps` trick mentioned in the issue description!

The intent of this PR is to make GPU-dependent derivations more "normal": they may request the `"cuda"` feature, which is no different from e.g. `"kvm"`, and that's all they request. Nix may or may not find builders configured to support that feature. A host running Nix within a FHS-distribution may be configured to support that feature as well, by mapping `libcuda.so` from a different location. A pure un-marked derivation may depend on the outputs of a GPU-dependent derivation. This, in particular, makes it easier to integrate GPU tests into existing Nix-based CIs like Hercules...

So, while you could update `makeImpureTest` to issue a different command that would e.g. account for "drivers from FHS-distributions", and while you could pass various extra `requiredSystemFeatures` markers to `makeImpureTest` as an argument, and while you cold probably eliminate the red-blue distinction, I feel like our CI needs are better described as "host configuration". That's then the new scope.

Madouura commented:
That makes sense.
This makes me wonder if we should get rid of ``makeImpureTest`` and/or merge it's wanted functions into this more generalized PR?
Having both risks the same problem C++ faces with it's multiple paradigms. ``makeImpureTest`` AFAIK isn't really used by any CI yet and there's only a couple packages in nixpkgs that actually use it.
For development, we're not running it on the CI anyway so if we can pass the system features we need in the derivation and just let nix take care of it, that would be ideal vs having to invoke the script and then manually executing it.
At least, with my limited knowledge that should be how it works. If not it should be trivial to just ignore ``makeImpureTest`` and manually specify the impure deps, or have a user-facing (developer) script to output the same commands the CI that picks up on the derivation's ``requiredSystemFeatures``.
In short, I think we can get rid of ``makeImpureTest`` and generalize this PR.

Madouura commented on nixos/modules/programs/nix-required-mounts.nix:
```diff
@@ -0,0 +1,89 @@
+    opengl.onFeatures = package.allowedPatterns.opengl.onFeatures;
+    opengl.paths = package.allowedPatterns.opengl.paths ++ driverPaths;
+    cuda.onFeatures = package.allowedPatterns.cuda.onFeatures;
+    cuda.paths = package.allowedPatterns.cuda.paths ++ driverPaths;
```
For instance, and it doesn't *have* to be done in this PR, but ROCm features/paths would go well here I believe.
[amdvlk/test.nix](https://github.com/NixOS/nixpkgs/blob/18e76671a16cebf6bcf52d9196d25a5a9f8c9ad7/pkgs/development/libraries/amdvlk/test.nix)
[clr/test-rocm-smi.nix](https://github.com/NixOS/nixpkgs/blob/18e76671a16cebf6bcf52d9196d25a5a9f8c9ad7/pkgs/development/rocm-modules/5/clr/test-rocm-smi.nix)

SomeoneSerge replied:
CC @Flakebi as the maintainer of amdvlk test

What do you think?

Flakebi replied:
First time I saw this. Are there any instructions on how to run such a test? (I suppose I need more than just `nix-build -A nixosTests.<pkg>.tests`?)

If this works as advertised, it sounds like we should get rid of the impureTests. Thanks for cc’ing me!

Madouura commented on pkgs/development/python-modules/torch/test-cuda.nix:
```diff
@@ -0,0 +1,22 @@
+{
+  nativeBuildInputs = [
+    (python.withPackages (_: [ torch ]))
+  ];
```
Seeing as we can use ``runCommand`` with this, it also seems to be possible to script together a test that can have acceptable failure conditions.
This is important to me because ``rsmitst`` from ``rocm-smi`` requires root, which we obviously do not want to give. I did test out a root-allowing variant with ``makeImpureTest``, but I don't prefer it.
I think the approach with this PR is superior, and more streamlined.

Madouura replied:
That would be ideal, since we could just mock root in the VM and not actually have to give the nix builder real root access to the host.

Madouura commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -225,6 +225,16 @@ stdenv.mkDerivation (finalAttrs: rec {
+      "blender-cuda-available"
+      {
+        nativeBuildInputs = [ finalAttrs.finalPackage ];
+        requiredSystemFeatures = [ "cuda" ];
```
Putting derivations in ``passthru`` can lead to splicing issues.
What if we compile on say ``aarch64`` but need to execute this on ``x86_64``?
Would ``preferLocalBuild`` alleviate this issue?

SomeoneSerge replied:
Oh, that's a very interesting question, I'll have to think about this. But I'm inclined to think this isn't a blocker, because do we even run tests non-native builds at this point? We disable checkPhases for python packages, and I'm not sure there's any automation in place to pick up the passthru tests from cross instantiations.

Pondering at the question though:

1. The semantics of the feature is that it asks for cuda devices to be available during the build, and driver usable on the build platform
2.If we instantiate `import nixpkgs { system = "aarch64-linux"; crossSystem = "x86_64-linux"; }` (I find the opposite direction more likely?) and try to access `blender.cudaAvailable`, we want
2.1 `blender`'s `buildPlatform` to be `aarch64-linux` and `x86_64-linux` the `hostPlatform`
2.2 `cudaAvailable`'s `buildPlatform` to be `x86_64-linux`
3. Then Nix will happily build `blender` on the `aarch64-linux`, and ask for a `x86_64-linux` remote builder for `cudaAvailable`. Or refuse to build if one's not available

SomeoneSerge replied:
Splicing aside, `pkgsHostTarget.callPackage ./test-gpu.nix { }` should do the trick

SomeoneSerge replied:
But how does splicing treat passthru? I've no idea

Madouura replied:
> But how does splicing treat passthru? I've no idea

This is the specific issue. https://github.com/NixOS/nixpkgs/issues/211340
``pkgsHostTarget.callPackage`` should work.
Are we able to define the ``buildPlatform`` for ``cudaAvailable``, or for that matter, for ``allowedPatterns``?

SomeoneSerge replied:
`allowedPatterns` are just data, there's nothing to build in there.

As for "defining the platform for `cudaAvailable`", isn't it exactly what we do if we `pkgs<X><Y>.callPackage`? Because that cross package set should contain its own `stdenv` with shifted build- and hostPlatforms. By the way, I was probably wrong about `pkgsHostTarget`, it should have been `pkgsTargetTarget` instead and I'm not sure if that actually works anyway:

```
❯ nix eval -f '<nixpkgs>' --arg crossSystem '{ config = "aarch64-unknown-linux-gnu"; }' pkgsHostTarget.buildPlatform.system
"x86_64-linux"
❯ nix eval -f '<nixpkgs>' --arg crossSystem '{ config = "aarch64-unknown-linux-gnu"; }' pkgsTargetTarget.buildPlatform.system
error: attribute 'buildPlatform' in selection path 'pkgsTargetTarget.buildPlatform.system' not found
```

Regardless, figuring out how to make passthru tests work in cross scenarios is way out of scope for this PR, there's the entire `python3Packages` full of passthroughs which still  doesn't care about non-native builds

SomeoneSerge commented:
There's definitely an overlap between this, `makeImpureTest`, and `__impureHostDeps` used on Darwin. All three are implemented in different layers: a pre-build-hook, a normal derivation + user interaction, and an actual special-case branch in Nix itself. There probably exists a solution that would address all three in a satisfying manner. Ideally, that solution would have to account for other possible implementations of Nix as well: so e.g. if we say pre-build-hooks are the way to go, then we need TVix to support pre-build-hooks and the extra-sandbox-paths command (well, whenever TVix implements builders). If we say that `__impureHostDeps` are to be treated by Nix specially, that also has to become a "standard" instead of an undocumented hack.

So let's keep it local for now and see how each works in the longer run. I think we can merge `__impureHostDeps` and `makeImpureTest` right away though (will require some patching either in Nix or in nixpkgs, to allow it outside darwin)

Madouura commented:
Overall, LGTM.
Haven't gotten around to testing yet, but I can't find anything blatantly wrong with the implementation.
I'll give it an approval once the passthru issue is addressed (or dismissed, since I'm unsure), and I test something.
Will probably just make a quick custom test using this with ``rocm-smi`` or something.
AFAIK ``allowedPatterns`` should allow me to do so.

Madouura reviewed (approved):
LGTM

samuela reviewed (commented):
"cuda" is an overloaded term in the sense that NVIDIA uses it to describe a user-land set of libraries but also distributes the ["NVIDIA driver"](https://www.nvidia.com/download/index.aspx) as part of CUDA Toolkit. 

IIUC this diff introduces the option to expose general GPU driver paths, eg `/run/opengl-driver/lib`, when running Nix tests. In that case is it appropriate to name the option "cuda"? For example, how would users enable GPU tests on AMD GPUs? Would that also be with the "cuda" option?

Madouura commented:
It might be more pertinent to generalize ``cuda`` and ``opengl`` into ``gpu``.
That way we can trivially add rocm (and eventually the intel equivalent) support.

SomeoneSerge commented:
Maybe rename `"cuda"` into `"nvidia"` in the allowed patterns, and add a similar preset for AMD GPUs? I don't think we want to expose more than is required, and when we test ROCm we probably don't need `/dev/nvidia*`

> For example, how would users enable GPU tests on AMD GPUs?

Just as for any other hypothetical co-processor, they'd locally set `programs.nix-required-mounts.allowedPatterns.amdgpu.paths = [ "/dev/whatsitcalled" ...]` and write derivations with `requiredSystemFeatures = [ "amdgpu" ]`. But they also won't have to, because we can add the default

SomeoneSerge commented:
I also just realized that the defaults for `"opengl"` are broken at least in case of NVidia, because libglvnd will probably eventually load libGL and libgbm from the NVidia driver, and some of those would open libnvidia-ml.so, and at least that one would try to access something like `/dev/nvidiactl`, I don't recall

samuela commented:
Thanks for explaining @SomeoneSerge ! I was under the impression that the only path necessary here was `/run/opengl-driver/lib`, but it sounds like `/dev/nvidia*` plays into this mix as well which I had not previously considered. 

I lean slightly away from "cuda" given that it is such an overloaded term, but happy to go with whichever naming you feel makes the most sense!

Flakebi commented:
> we could also try exposing them inside VMs inside Nix sandboxes using PCI passthrough

As this comes up from time to time, PCI passthrough is tempting and would make lots of things possible. But GPU passthrough requires careful setup of hardware, bios and software and does not reliably work (on various hardware), so I think this is not a feasible way for tests.

If it gets more usable in the future, it surely is worth exploration, but currently we are far away from that.

SomeoneSerge committed:
programs.nix-required-mounts: presets.cuda -> nvidia-gpu

This hopefully clarifies that the preset configures the hook to expose
"nvidia devices", which includse both the userspace driver and the
device nodes.

The derivations  still declare requiredSystemFeatures = [ "cuda" ] to
explicitly indicate they need to use the CUDA functionality and expect a
libcuda.so and a CUDA-capable device. Ideally, we'd also include the
specific CUDA architectures (sm_86, etc) in feature names.

Derivations that use a co-processor but do not care about the vendor or
even the particular interface may ask for the more generic "opengl",
"vulkan", or "gpu" features. It is then responsibility of the host
declaring the support for this feature to ensure the drivers and
hardware are appropriately set up.

SomeoneSerge committed:
nix-required-mounts: guest and host paths may differ

SomeoneSerge committed:
nix-required-mounts: restore the followSymlinks option

This way pkgs.nix-required-mounts is "correct" even before
we override it in the NixOS module

SomeoneSerge committed:
nix-required-mounts: restore (optional) symlink support

SomeoneSerge commented:
In the last push:
- droppe dthe "opengl" preset because I actually couldn't guarantee I'm mounting all of the required device nodes (say, nvidia ships a custom libgl, but it could be trying to open some device nodes other than /dev/video* because, wlellll, it's custom innt)
- renamed the presetn into "nvidia-gpu"
- the derivations still ask for what I'd call "interfaces" that the host needs to support, e.g. cuda (libcuda and the device nodes), vulkan (libvulkan and the devices), or opengl (libglx, libegl, libgbm and the devices) - as opposed to asking for vendors. It's the responsibility of the host declaring a feature support to actually provide the required software and hardware stack, and to mounts things properly in the sandbox
- restored the symlink support, now optional, which was required to make the non-overridden `pkgs.nix-required-mounts` "valid"; the nixos module still overrides the package, and disables the mostly untested symlink support
- restored the option of non-matching host/guest paths, which might be important for fhs distributions (you could mount driver libraries from /usr/lib/ to /run/opengl-driver/lib in the sandbox)

What I haven't done but I think would be appropriate is rewrite the defaults to only expose parts of `/run/opengl-driver/lib` instead of the whole directory

SomeoneSerge commented:
Upd: switched the workstation to HEAD and rerun the passthru tests. They pass. I may be incapacitated at the present moment to judge about whether the last changes make any sense, but the tests do pass

tomberek commented:
Random comment: I believe __impureHostDeps will gracefully degrade if the requested paths do not exist.

SomeoneSerge commented on nixos/modules/programs/nix-required-mounts.nix:
```diff
@@ -0,0 +1,96 @@
+      '';
+    });
+
+  driverPaths = [
+    # symlinks in /run/opengl-driver/lib:
+    pkgs.addOpenGLRunpath.driverLink
+
+    # the actual userspace drivers in /nix/store:
+    config.hardware.opengl.package
+  ] ++ config.hardware.opengl.extraPackages;
```
It just occurred to me that I'm not explicitly mounting targets of the RUNPATHs in the userspace drivers. For instance, I cannot explain why the exact `libc` version linked to by `libcuda.so` happens to be available in the sandbox, but I suspect that's just a coincidence

Kiskae replied:
You probably need to use https://nixos.org/manual/nix/stable/language/advanced-attributes#adv-attr-exportReferencesGraph to get the complete closure of the driver runtime.

SomeoneSerge committed:
nix-required-mounts: allow overriding the rendered config

SomeoneSerge committed:
nixos/nix-required-mounts: mount the runtime closures

SomeoneSerge committed:
nix-required-mounts: enforce that host paths exist

SomeoneSerge committed:
nix-required-mounts: print -> logging

SomeoneSerge committed:
nixos/nix-required-mounts: allow passing extra arguments to the hook

SomeoneSerge committed:
cudaPackages.saxpy: passthru: test gpu/runtime

SomeoneSerge commented:
In the latest push:

- Added an option to use `exportReferencesGraph` to expand `paths` into their closures, as suggested by @Kiskae. No automated tests. If I add any, I probably do so in a separate PR
- Added checks that the user-specified patterns do resolve into some paths in the host filesystem. Didn't add tests though, and also I do not run the checks at system activation time, so one could in principle switch into a configuration with the broken hook and then I guess they have to delete `/etc/nix/nix.conf` and do something manually. I'm not planning to address that in this PR

The rest is just stuff moving around.
Waiting for OfBorg

SomeoneSerge commented:
Does Ofborg build individual `passthru.tests` attributes separately, or does it try to e.g. assemble them in a `buildEnv`? Because in the latter case a derivation with `requiredSystemFeatures = [ "cuda" ]` would halt all of the tests, including those that Ofborg would normally build

SomeoneSerge commented:
Rebased on master (after the jetson PR), but I guess I still should verify if Ofborg is OK. I missed it the last time it reported any status other than merge conflict

SomeoneSerge committed:
cudaPackages: move cuda tests from passthru.tests

Otherwise we crash Ofborg

SomeoneSerge committed:
python3Packages.torch.gpuChecks: add rocm

SomeoneSerge committed:
blender.gpuChecks: add unwrapped

An unwrapped check for `nix run`-ing on the host platform,
instead of `nix build`-ing in the sandbox

E.g.:

```
❯ nix run -f ./. --arg config '{ cudaSupport = true; cudaCapabilities = [ "8.6" ]; cudaEnableForwardCompat = false; allowUnfree = true; }' -L blender.gpuChecks.cudaAvailable.unwrapped
Blender 4.0.1
Read prefs: "/home/user/.config/blender/4.0/config/userpref.blend"
CUDA is available
Blender quit
❯ nix build -f ./. --arg config '{ cudaSupport = true; cudaCapabilities = [ "8.6" ]; cudaEnableForwardCompat = false; allowUnfree = true; }' -L blender.gpuChecks
blender> Blender 4.0.1
blender> could not get a list of mounted file-systems
blender> CUDA is available
blender> Blender quit
```

SomeoneSerge commented:
I've only recently learned about `pkgs/build-support/references-by-popularity/closure-graph.py` and `pkgs/build-support/closure-info.nix` which are great references for using `exportReferencesGraph` I wish I knew about before I wrote my ugly script:)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/on-nixpkgs-and-the-ai-follow-up-to-2023-nix-developer-dialogues/37087/2

SomeoneSerge commented:
@GaetanLepage and I just tested this on one of his NixOS hosts. We experimented with optionally enabling the OpenCL and CUDA tests that require access to a coprocessor inside the sandbox.

- [ ] We found a few rough edges that I want to fix before the merge (e.g. I shouldn't require `/dev/video*` to exist for the nvidia stuff)
- [ ] I also noticed I didn't add any documentation or release notes - I'll fix that too.

A few shortcomings remain that I'm aware of and consider minor and out of scope:

- [ ] This a fairly ad hoc scheme oblivious of the "usual" ways people used to use to configure nvidia devices in containers, e.g of the `CUDA_VISIBLE_DEVICES` env variable. This is not a problem, my plan is to just add an integration with CDI (https://github.com/NixOS/nixpkgs/pull/284507) later. I think this won't require changing any of the interfaces
- [ ] The runtime closure computation is... I'm not proud of this implementation. IMO this is OK because (1) we tested the current implementation, (2) I can simplify the implementation in a subsequent PR without changing the interfaces.
- [ ] We're discussing alternative naming schemes for gpu tests in the [#cuda:nixos.org](https://matrix.to/#/%23cuda%3Anixos.org/%24Uf81HxHEutaiFClu7LOCY-FiWg6DkO-KPxSlTP-inWQ?via=someonex.net&via=matrix.org&via=nixos.dev&via=amber.tel) matrix room. We need a convention so that we can build CI around it. The reason I'm not happy with arbitrary name is that it's beneficial to expose the `nix build`-able "tests" (currently e.g. `gpuChecks.cuda`) and `nix run`-able "tester scripts" (currently e.g. `gpuChecks.cuda.unwrapped`). There are cases where it's easy to write one but not the other. More in the matrix room

**I intend to address the first two items and merge this PR by next Monday. Objections?**

SomeoneSerge committed:
nix-required-mounts: cuda: /dev/video* may not exist and aren't relevant

SomeoneSerge committed:
nix-required-mounts: nixfmt

SomeoneSerge committed:
nix-required-mounts: refactor: drop unused arguments

SomeoneSerge committed:
cudaPackages: updated convention for gpu/runtime checks

Runtime tests (derivations asking for a relaxed sandbox) are now
expected at p.gpuCheck, p.gpuChecks.<name>, or at
p.tests.<name>.gpuCheck.

SomeoneSerge commented:
> ...merge this PR by next Monday

Wrong Monday. I now fixed the `/dev/video*` issue and added release notes. I also changed slightly the paths for these "gpu tests" (cf the commit message for https://github.com/NixOS/nixpkgs/pull/256230/commits/d6b3abcb87e64c6d5cce79c22eca3bdccd160fd3). The `exportReferencesGraph` logic can be rewritten later as part of the effort for making nvidia's [CDI integration work](https://github.com/NixOS/nixpkgs/issues/290609#exportReferencesGraph).

Waiting for ofborg and merging in the morning.

GaetanLepage commented:
Result of `nixpkgs-review pr 256230` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>nix-required-mounts</li>
    <li>nix-required-mounts.dist</li>
  </ul>
</details>

SomeoneSerge commented:
This time around Ofborg dismisses `passthru.tests` (in bulk) based on licenses rather than `requiredSystemFeatures`. There are precedents for this already in Nixpkgs, e.g. `python3Packages.jax.tests`, so I'm not going to delay anymore. Properly resolving this depends either on https://github.com/NixOS/ofborg/issues/678 or on https://github.com/NixOS/ofborg/pull/644

```
❯ nix build --impure .#blender.tests.tester-cudaAvailable.gpuCheck -L # --impure is for the global allowUnfreePredicate
test-cuda> Blender 4.1.1
test-cuda> could not get a list of mounted file-systems
test-cuda> CUDA is available
test-cuda> Blender quit
```

🚅 🚄 🚄

trofi commented:
In some locations `test-cuda.nix` is not present, fails the eval as:

```
$ nix build --no-link -f. python3Packages.pytorch-bin.gpuChecks
error:
       … while evaluating a branch condition
         at lib/customisation.nix:263:8:
          262|
          263|     in if missingArgs == {}
             |        ^
          264|        then makeOverridable f allArgs

       … while calling the 'listToAttrs' builtin
         at lib/attrsets.nix:647:5:
          646|     set:
          647|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |     ^
          648|

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: path 'pkgs/development/python-modules/torch/test-cuda.nix' does not exist
```
